lib/api/apiUtils/apiCallers/callPostObject.js

```diff
@@ -0,0 +1,210 @@
+const { auth, errors } = require('arsenal');
+const busboy = require('@fastify/busboy');
+const writeContinue = require('../../../utilities/writeContinue');
+const fs = require('fs');
+const path = require('path');
+const os = require('os');
+
+/** @see doc: https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-HTTPPOSTForms.html#HTTPPOSTFormDeclaration */
+const MAX_FIELD_SIZE = 20 * 1024; // 20KB
+/** @see doc: https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html */
+const MAX_KEY_SIZE = 1024;
+const POST_OBJECT_OPTIONAL_FIELDS = [
+    'acl',
+    'awsaccesskeyid',
+    'bucket',
+    'cache-control',
+    'content-disposition',
+    'content-encoding',
+    'content-type',
+    'expires',
+    'policy',
+    'redirect',
+    'tagging',
+    'success_action_redirect',
+    'success_action_status',
+    'x-amz-meta-',
+    'x-amz-storage-class',
+    'x-amz-security-token',
+    'x-amz-signgnature',
+    'x-amz-website-redirect-location',
+];
+
+async function authenticateRequest(request, requestContexts, log) {
+    return new Promise(resolve => {
+        // TODO RING-45960 remove ignore auth check for POST object here
+        auth.server.doAuth(request, log, (err, userInfo, authorizationResults, streamingV4Params) =>
+            resolve({ userInfo, authorizationResults, streamingV4Params }), 's3', requestContexts);
+    });
+}
+
+async function parseFormData(request, response, requestContexts, log) {
+    /* eslint-disable no-param-reassign */
+    let formDataParser;
+    try {
+        formDataParser = busboy({ headers: request.headers });
```

Comment:
Some more details about `multipart/form-data`, parsing by `busboy` and aws behavior with some links to documentation:

<details>
<summary>Content-Type</summary>

Documentation for the Content-Type, with the boundary directive and example with body: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type

Again, `busboy` will accept `application/x-www-form-urlencoded` (https://github.com/fastify/busboy/blob/master/lib/types/urlencoded.js#L9)

That will unnecessarily read the whole body and trigger an error `400 InvalidArgument` instead of the `412 Precondition Failed` returned by aws and stopping the request early.
</details>

<details>
<summary>Content-Disposition</summary>

The documentation about `Content-Disposition` gives you details for its usage in `multipart/form-data`: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition

In the body, a part will look like

```
--boundary
Content-Disposition: form-data; name="f1"

value1
--boundary
```

Here `busboy` will parse this and emit a `field` event with `fieldname` being `f1`

You can send this with:
- an html form with an `<input type="text" name="f1">`
- curl (https://curl.se/docs/manpage.html#-F)
    - `curl -F f1=value1`
    - `curl -F "f1=</etc/hostname"` (this put the content of the file but not the filename)

You can also have

```
--boundary
Content-Disposition: form-data; name="f2"; filename="example.txt"

value2
--boundary
```

You can send this with:
- an html form with an `<input type="file" name="f2">`
- curl (https://curl.se/docs/manpage.html#-F)
    - `curl -F "f2=@example.txt"` (filename will be example.txt)
    - `curl -F "f2=value2;filename=example.txt"

Notice the `filename` directive, this will make `busboy` emit a `file` event with `fieldname`  being `f2`.
As described in busboy documentation: https://www.npmjs.com/package/@fastify/busboy#busboy-methods with the default config `isPartAFile`.

You can notice in AWS documentation about the `file` form field: https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPOST.html#RESTObjectPOST-requests-form-fields

> The file or text content.
> ...
> Type: File or text content

So you can either send
`Content-Disposition: form-data; name="file"` (with an input of type text)
or
`Content-Disposition: form-data; name="file"; filename="whatever.txt"`

Which means by default the `file` form field can be emitted either by the `field` or `file` busboy event.

Also by testing on AWS, it accepts other form fields like `key`, `acl` having a `filename` directive (meaning coming from an input type="file"), example this should be possible:

- `Content-Disposition: form-data; name="key"; filename="whatever.txt"`
- `Content-Disposition: form-data; name="acl"; filename="whatever.txt"`

So any form field can be emitted by either the `field` or `file` busboy event.
The current processing / validation not valid, it should be performed based on the `fieldname` and not the busboy event.

Note that it could be interesting to replace the default behavior of busboy and defining `isPartAFile` to consider a file only the `fieldname === 'file'` (with potentially presence of `filename`).
___

When doing that there should also be some limits defined on busboy, if a form field other than `file` is being a big file and now parsed as a string instead of a stream, limit would help busboy to stop reading it. (The default `limit.fieldSize` is 1MiB but AWS accepts only 20KB for the form-data exluding file).

> The form data and boundaries (excluding the contents of the file) cannot exceed 20KB.
Note: it looks like it's not just the field value but potentially the whole boundary with `Content-Disposition` and potential headers of the part, but I'm not sure, might need to be tested.


</details>

<details>
<summary>${filename} replacement</summary>

About `${filename}` replacement, it's not only on the key, but all form fields: https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-HTTPPOSTForms.html

> The variable ${filename} is automatically replaced with the name of the file provided by the user and is recognized by **all form fields**. If the browser or client provides a full or partial path to the file, only the text following the last slash (/) or backslash (\) is used (for example, C:\Program Files\directory1\file.txt is interpreted as file.txt). If no file or file name is provided, the variable is replaced with an empty string.


Also the documentation defines  how `filename` should be sanitized (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition#filename)
Busboy should sanitize it, but you could add some tests with filename containing `../xxx`, `a/b/c`, etc to ensure the `${filename}` works as aws.

And validation on form fields, like `key` length, should then be performed after `${filename}` replacement as it could impact the value.

</details>

<details>
<summary>Additional validation</summary>

You can receive a `Content-Disposition: form-data` without name, that would make an undefined `fieldname` and crash:
`TypeError: Cannot read properties of undefined (reading 'toLowerCase')`

Also by testing on AWS, they seem to trim the `fieldname` of whitespaces: `name="  \n    key   \n    ` will should match `key`.
___

It's also possible to have multiple parts with the same `name`.
For example with this:

```html
<input type="text" name="key">
<input type="text" name="key">
<input type="text" name="acl">
<input type="text" name="acl">
<input type="file" name="file" multiple>
```

or `curl -F f1=value1 -F f1=value2`

We should ensure, just like AWS that we don't have multiple times the same form field, whether busboy emits it via `field` or `file`.
___

Maybe the file should not be read if the parsing failed and an error was caught.
And when reading (pipe) the file stream, we need to catch error, otherwise it can crash:
`Error: Part terminated early due to unexpected end of multipart data`

It seems the `finish` event is triggered once the file is completely read, and there is a validation logic in there, the logic could happen earlier before writting file to disk.

___

There is a validation on `Content-Length as well:

```
    <Code>EntityTooLarge</Code>
    <Message>Your proposed upload exceeds the maximum allowed size</Message>
    <ProposedSize>6442453100</ProposedSize>
    <MaxSizeAllowed>5368730624</MaxSizeAllowed>
```
</details>

<details>
<summary>
You should find attached a javascript script to write the raw body form-data easier than using `nc`.
It allows to test invalid bodies that are usually well formated by lib like `form-data` or by `curl` or `postman`.

Also you can add a delay between writes to test server timeout behavior.
</summary>

```javascript
const http = require('http');

const CRLF = '\r\n';

const body = `\
--test
Content-Disposition: form-data; name="key"

some_key
--test
Content-Disposition: form-data; name="file"; filename="some_name"

a
--test--
`.replace(/\n/g, CRLF);

const options = {
    method: 'POST',
    headers: {
        'Host': 'micktestpost.localhost',
        'Content-Type': 'multipart/form-data; boundary=test',
        'Content-Length': Buffer.byteLength(body),
    },
};

const req = http.request('http://localhost:8000', options, (res) => {
    console.log(req)
    console.log('RESPONSE', { statusCode: res.statusCode, message: res.statusMessage, headers: res.headers });

    res.on('data', (chunk) => {
        console.log(`BODY: ${chunk}`);
    });
    res.on('end', () => {
        console.log('END');
    });
});

req.on('error', (e) => {
    console.error(`problem with request: ${e.toString()}`);
});

// Split body in block of 20 chars
const split = body.match(/[\s\S]{1,20}/g);
// Increase delay without writting to test server timeout
const writeDelayMs = 1_000;

function writeBodyPart(i) {
    if (i < split.length) {
        console.log('Writting', i, split[i])
        req.write(split[i]);
        setTimeout(() => writeBodyPart(i + 1), writeDelayMs);
    } else {
        console.log('Done writting body')
        req.end();
    }
}

// setTimeout(() => writeBodyPart(0), writeDelayMs)

req.write(body);
req.end();
```
</details>

Generating the body raw without a lib can help you test any kind of invalid body to make sure you use busboy correctly and validate all input before using them to prevent a crash of the server.

Busboy already takes care of all the parsing, we need to configure it correctly (cf busboy readme) to match AWS spec and validate inputs to prevent a crash.

Ideally you could extract the parsing / validation (parseFormData function) with busboy into its own isolated class to make it easier for testing. And if you want to be consistent with callbacks usage instead of promises and async await, you can pass a callback as input of your parsing function to call once parsing /validation is done.